src/main/java/sparksoniq/jsoniq/runtime/iterator/operational/CastIterator.java

```diff
@@ -0,0 +1,107 @@
+import sparksoniq.semantics.types.AtomicTypes;
+import sparksoniq.semantics.types.ItemTypes;
+import sparksoniq.semantics.types.SingleType;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+public class CastIterator extends UnaryOperationIterator {
+    private static final long serialVersionUID = 1L;
+    private final SingleType _singleType;
+
+    private static final Map<String, Class> PRIMITIVE_NAME_TYPE_MAP = new HashMap<>();
+
+    static {
+        PRIMITIVE_NAME_TYPE_MAP.put("JSONItem", JsonItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("ObjectItem", ObjectItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("ArrayItem", ArrayItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("AtomicItem", AtomicItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("StringItem", StringItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("IntegerItem", IntegerItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("DecimalItem", DecimalItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("DoubleItem", DoubleItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("BooleanItem", BooleanItem.class);
+        PRIMITIVE_NAME_TYPE_MAP.put("NullItem", NullItem.class);
+    }
+
+    public CastIterator(RuntimeIterator child, SingleType singleType, IteratorMetadata iteratorMetadata) {
+        super(child, OperationalExpressionBase.Operator.CAST, iteratorMetadata);
+        this._singleType = singleType;
+    }
+
+    @Override
+    public Item next() {
+        if (this._hasNext) {
+            String targetType = ItemTypes.getItemTypeName(_singleType.getType().toString());
+
+            List<Item> items = new ArrayList<>();
+            _child.open(_currentDynamicContext);
+            while (_child.hasNext()) {
+                items.add(_child.next());
+                if (items.size() > 1) {
+                    _child.close();
+                    this._hasNext = false;
+                    throw new UnexpectedTypeException(" Sequence of more than one item can not be treated as type "
+                            + targetType, getMetadata());
+                }
+            }
+            _child.close();
+            this._hasNext = false;
+
+            Item item = items.get(0);
+            String itemType = ItemTypes.getItemTypeName(item.getClass().getSimpleName());
+
+            if (itemType.equals(targetType)) return item;
+
+            String message = String.format("\"%s\": value of type %s is not castable to type %s",
+                    item.serialize(), itemType, targetType);
+
+            AtomicItem atomicItem = CastableIterator.checkInvalidCastable(item, getMetadata(), _singleType);
+
+            Class classToLoad = PRIMITIVE_NAME_TYPE_MAP.get(_singleType.getType().toString());
+            assert classToLoad != null;
+
+            if (atomicItem.isCastableAs(_singleType)) {
+                try {
+                    return atomicItem.castAs((AtomicItem) classToLoad.newInstance());
```

Comment:
Creating new instances should be avoided for performance reasons. This code will potentially be executed billions of times.

Passing the SingleType instance will avoid this.